api/crossmodelrelations/crossmodelrelations.go

```diff
@@ -53,6 +66,11 @@ func (c *Client) PublishRelationChange(change params.RemoteRelationChangeEvent)
 	args := params.RemoteRelationsChanges{
 		Changes: []params.RemoteRelationChangeEvent{change},
 	}
+	// Use any previously cached discharge macaroons.
+	if ms, ok := c.cache.Get(change.RelationToken); ok {
```

Comment:
It's sorted of repeated but different. Different arg types and return value types etc so logic is slightly different. Some places it's the same. But the Go philosophy tends to be simplicity even if that means extra boilerplate. Trying to extract a closure would work for some of the methods but not all. So we would be adding cognitive overhead there. I erred on the side of simple to read code.  I'll take another look to see if something simple and readable can be done.